### PR TITLE
[New Control] - GR7 | Validation 3 | Certificate Validity: Application Gateway (M)

### PR DESCRIPTION
**Pull request by singhgss**
## Overview/Summary

This Pull Request adds a new control for ensuring the validity of SSL certificates in the Azure Application Gateway listeners. It includes checks for certificate expiration and verification that certificates are from an approved certificate authority as determined by the department.

## This PR fixes/adds/changes/removes

1. Adds control to validate SSL certificates in Azure Application Gateway listeners.
2. Checks that certificates are current and not expired.
3. Verifies that certificates are from a department-approved certificate authority (CA) or that the department is using a well-known CA or Key Vault for certificate management.

### Breaking Changes

1. N/A

## Testing Evidence

The following steps were taken to test the functionality of the new control:

1. Checked if an Azure Application Gateway is being used by the department. If not, the control marks as non-compliant or not applicable.
2. Verified if any SSL certificates are attached to a listener. If no SSL certificates are found, the control marks as non-compliant.
3. Ensured that certificate expiration dates are greater than the current date to avoid using expired certificates.
4. Confirmed that certificates are from an approved certificate authority by verifying one of the following:
   - The "well-known" CA toggle is enabled.
   - An allow-list of approved certificates exists.
   - The department is using Azure Key Vault to store and manage approved certificates.
   
Testing results:

- Certificates were validated for expiration.
- Certificates were successfully checked against the approved CA list.

image

image

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/issues)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/tree/main)
- [x] Performed testing and provided evidence.
- [ ] Updated relevant and associated documentation.
- [x] Ensured PowerShell module versions have been updated (manually or with the ./tools/Update-ModuleVersions.ps1 script)

### CODE REVIEW

File: src/GUARDRAIL 7 PROTECTION OF DATA-IN-TRANSIT/Audit/Check-ApplicationGatewayCertificateValidity.psm1
Comment (dutt0) [inline src/GUARDRAIL 7 PROTECTION OF DATA-IN-TRANSIT/Audit/Check-ApplicationGatewayCertificateValidity.psm1:1]: $PsObject and $moduleOutput code is duplicated. Can we put line 93-200 inside 'if' and remove the duplicates?

File: src/Guardrails-Common/GR-Common.psm1
Comment (dutt0) [inline src/Guardrails-Common/GR-Common.psm1:1]: [OutputType([int])] is twice in the file

File: src/GUARDRAIL 2 MANAGE ACCESS/Audit/Check-OnlineAttackCountermeasures.psm1
Comment (dutt0) [inline src/GUARDRAIL 2 MANAGE ACCESS/Audit/Check-OnlineAttackCountermeasures.psm1:1]: need to update the zip module for this GR2 control

File: src/GUARDRAIL 7 PROTECTION OF DATA-IN-TRANSIT/Audit/Check-ApplicationGatewayCertificateValidity.psm1
Comment (singhgss) [reply src/GUARDRAIL 7 PROTECTION OF DATA-IN-TRANSIT/Audit/Check-ApplicationGatewayCertificateValidity.psm1:1]: @dutt0  Thats just so we dont go further and return from the point where document is not found to compare CA list.

**Review by dutt0 — APPROVED** [review]

**Pull request merged**